modules/apps/asset/asset-taglib/src/main/resources/META-INF/resources/asset_categories_selector/AssetVocabularyCategoriesSelector.es.js

```diff
@@ -12,300 +12,276 @@
-		const sub = (str, obj) => str.replace(/\{([^}]+)\}/g, (_, m) => obj[m]);
-
-		const uri = sub(decodeURIComponent(this.portletURL), {
-			selectedCategories: this.selectedItems
-				.map(item => item.value)
-				.join(),
-			singleSelect: this.singleSelect,
-			vocabularyIds: this.vocabularyIds.concat()
-		});
+	useEffect(() => {
+		ref.current = value;
+	}, [value]);
 
-		const itemSelectorDialog = new ItemSelectorDialog({
-			eventName: this.eventName,
-			title: Liferay.Language.get('select-categories'),
-			url: uri
-		});
+	return ref.current;
+}
 
-		itemSelectorDialog.open();
-		itemSelectorDialog.on('selectedItemChange', event => {
-			const selectedItems = event.selectedItem;
-
-			if (selectedItems) {
-				this.selectedItems = Object.keys(selectedItems).map(itemKey => {
-					return {
-						label: selectedItems[itemKey].value,
-						value: selectedItems[itemKey].categoryId
-					};
-				});
-			}
-		});
-	}
-
-	_handleInputFocus(event) {
-		this.emit('inputFocus', event);
-	}
-
-	/**
-	 * Converts the list of selected categories into a comma-separated serialized
-	 * version to be used as a fallback for old services and implementations.
-	 *
-	 * @private
-	 * @return {string} The serialized, comma-separated version of the selected items.
-	 */
-	_getCategoryIds() {
-		return this.selectedItems
-			.map(selectedItem => selectedItem.value)
-			.join();
-	}
-
-	/**
-	 * Shows the category error.
-	 *
-	 * @private
-	 */
-	_handleErrorAddinglabelItem(event) {
-		if (event.data.itemDoesNotExists) {
-			this._typedCategory = event.target.inputValue;
-			this._unexistingCategoryError = true;
+function AssetVocabulariesCategoriesSelector({
+	eventName,
+	id,
+	isValid = true,
+	groupIds = [],
+	inputName,
+	label,
+	onSelectedItemsChange = () => {},
+	portletURL,
+	required,
+	selectedItems = [],
+	singleSelect,
+	sourceItemsVocabularyIds = [],
+	useFallbackInput
+}) {
+	const [inputValue, setInputValue] = useState('');
+	const [invalidItems, setInvalidItems] = useState([]);
+
+	const {refetch, resource} = useResource({
+		fetchOptions: {
+			body: new URLSearchParams({
+				cmd: JSON.stringify({
+					'/assetcategory/search': {
+						'-obc': null,
+						end: 20,
+						groupIds,
+						name: `%${inputValue.toLowerCase()}%`,
+						start: 0,
+						vocabularyIds: sourceItemsVocabularyIds
+					}
+				}),
+				p_auth: Liferay.authToken
+			}),
+			credentials: 'include',
+			method: 'POST',
+			'x-csrf-token': Liferay.authToken
+		},
+		link: `${window.location.origin}${themeDisplay.getPathContext()}
+				/api/jsonws/invoke`
+	});
+
+	const previousInputValue = usePrevious(inputValue);
+
+	useEffect(() => {
+		if (inputValue && inputValue !== previousInputValue) {
+			refetch();
 		}
-	}
+	}, [inputValue, previousInputValue, refetch]);
+
+	const getUnique = (arr, compKey) => {
+		return arr
+			.map(e => e[compKey])
+			.map((e, i, final) => final.indexOf(e) === i && i)
+			.filter(e => arr[e])
+			.map(e => arr[e]);
```

Comment:
Could we use non-abbreviated names here? What is `e`? What is `compKey`? And `final` is a final what?

Reply:
:+1: for non-abbreviated names

As you know _naming is difficult_, and _changing_ that naming when you didn't write it makes it a bit more fun ;) 

I guess `e` is the "element", `compKey` is the property used to check the "uniqueness" of the element and `final` is the array map was called upon, so I'll rename these arguments and once everything is ready (I'm going to re-order and re-send this PR) we can keep on discussing it there - @carloslancha will also correct me if my assumptions are wrong.